File: job-scraper-backend/scrapers/greatrwandajobs_scraper.py
# scrapers/greatrwandajobs_scraper.py

import requests
import re
from bs4 import BeautifulSoup
from .base_scraper import BaseScraper
from .keywords import DEFAULT_KEYWORDS

class GreatRwandaJobsScraper(BaseScraper):
    """A scraper for 'greatrwandajobs.com' job listings."""

    def _fetch_categories_from_website(self):
        """
        Dynamically fetches all job categories from the website's category dropdown.
        Returns a list of dictionaries with 'value' and 'name' keys.
        """
        HEADERS = {
            'User-Agent': 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/91.0.4472.124 Safari/537.36'
        }
        
        try:
            # Fetch the main page to get the category dropdown
            response = requests.get("https://www.greatrwandajobs.com/jobs/", headers=HEADERS, timeout=20)
            response.raise_for_status()
            
            soup = BeautifulSoup(response.content, "html.parser")
            
            # Find the category select element
            category_select = soup.find("select", {"id": "category", "name": "category[]"})
            
            if not category_select:
                print("Warning: Category select element not found on webpage")
                return []
            
            categories = []
            options = category_select.find_all("option")
            
            for option in options:
                value = option.get("value")
                name = option.get_text(strip=True)
                
                # Skip empty options or options without value
                if value and name and value != "":
                    categories.append({
                        "value": value,
                        "name": name
                    })
            
            print(f"Successfully fetched {len(categories)} categories from website")
            return categories
            
        except requests.RequestException as e:
            print(f"Error fetching categories from website: {e}")
            return []
        except Exception as e:
            print(f"Error parsing categories from website: {e}")
            return []

    def _get_category_urls(self):
        """
        Constructs URLs for job categories that match the default IT-related keywords
        by dynamically fetching categories from the website.
        """
        # Fetch categories dynamically from the website
        all_categories = self._fetch_categories_from_website()
        
        # Fallback to static categories if website fetch fails
        if not all_categories:
            print("Using fallback static categories")
            all_categories = [
                {"value": "10", "name": "Engineering jobs in Rwanda"},
                {"value": "13", "name": "Computer/ IT jobs in Rwanda"},
                {"value": "47", "name": "Data, Monitoring, and Research jobs in Rwanda"},
                {"value": "52", "name": "Technician jobs in Rwanda"}
            ]
        
        keyword_pattern = r'\b(' + '|'.join(DEFAULT_KEYWORDS) + r')\b'
        urls = []

        for category in all_categories:
            category_name = category["name"]
            category_id = category["value"]

            # Check if any default keyword is in the category name
            if re.search(keyword_pattern, category_name, re.IGNORECASE):
                # Format category name for URL:
                # 1. Remove " jobs in Rwanda" suffix
                # 2. Make it lowercase
                # 3. Remove special characters except spaces, hyphens, and slashes
                # 4. Replace spaces and slashes with hyphens
                formatted_name = category_name.lower()
                # Remove special characters and replace spaces/slashes with hyphens
                formatted_name = re.sub(r'[^a-z0-9\s/-]', '', formatted_name)
                formatted_name = re.sub(r'[\s/]+', '-', formatted_name)
                # Remove any trailing hyphens
                formatted_name = formatted_name.strip('-')
                
                url = f"https://www.greatrwandajobs.com/job-categories/newest-jobs/category-{formatted_name}-{category_id}"
                urls.append(url)
        
        print(f"Found {len(urls)} relevant category URLs: {urls}")
        
        # If no relevant categories were found, use the default IT jobs page as a fallback
        if not urls:
            urls.append("https://www.greatrwandajobs.com/job-categories/newest-jobs/category-computer-it-jobs-in-rwanda-13")
        
        return urls

    def scrape(self, keyword=None):
        BASE_URL = "https://www.greatrwandajobs.com"
        HEADERS = {
            'User-Agent': 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/91.0.4472.124 Safari/537.36'
        }

        # Define keywords for filtering relevant jobs
        all_jobs = []
        company_names = set()
        
        if keyword:
            # If a specific keyword is provided, search the main jobs page.
            search_url = f"https://www.greatrwandajobs.com/jobs/?search_keywords={keyword}"
            urls_to_scrape = [search_url]
        else:
            # Otherwise, get the list of IT-related category URLs
            urls_to_scrape = self._get_category_urls()
            if not urls_to_scrape:
                # Fallback to a default URL if no categories are found
                urls_to_scrape = ["https://www.greatrwandajobs.com/job-categories/newest-jobs/category-computer-it-jobs-in-rwanda-13"]

        for URL in urls_to_scrape:
            print(f"Scraping URL: {URL}")
            try:
                page = requests.get(URL, headers=HEADERS, timeout=20)
                page.raise_for_status()
            except requests.RequestException as e:
                print(f"Error fetching the URL {URL}: {e}")
                continue # Move to the next URL

            soup = BeautifulSoup(page.content, "html.parser")
            # Find all job containers - each job is wrapped in its own js-jobs-wrapper div
            job_containers = soup.find_all("div", id="js-jobs-wrapper")
            print(f"Found {len(job_containers)} job containers on {URL}")
            
            job_elements = []
            for job_container in job_containers:
                # Find the job posting within each container
                toprow = job_container.find("div", class_="js-toprow")
                if toprow:
                    job_elements.append(toprow)
            
            print(f"Found {len(job_elements)} job elements on {URL}")

            for job_element in job_elements:
                title_element = job_element.find("a", class_="jobtitle")
                if not title_element:
                    print("Skipping job element: no title found")
                    continue
                    
                title = title_element.get_text(strip=True)
                link = BASE_URL + title_element["href"]
                
                # Fix company extraction to handle missing elements
                company = "N/A"
                company_div = job_element.find("div", class_="js-image")
                if company_div:
                    company_img = company_div.find("img")
                    if company_img:
                        company = company_img.get("title", "N/A")
                
                details_container = job_element.find("div", class_="js-second-row")
                category, posted_date, deadline_date, duty_station = "N/A", "N/A", "N/A", "N/A"

                if details_container:
                    category_element = details_container.find("span", string="Job Category: ")
                    if category_element and category_element.parent:
                        category = category_element.find_next_sibling(string=True).strip()

                    posted_element = details_container.find("span", string="Posted: ")
                    if posted_element and posted_element.parent:
                        posted_date = posted_element.find_next_sibling(string=True).strip()

                    deadline_element = details_container.find("span", string=re.compile(r"Deadline of this Job"))
                    if deadline_element and deadline_element.parent:
                        deadline_text = deadline_element.parent.get_text(strip=True)
                        if ":" in deadline_text:
                            deadline_date = deadline_text.split(":", 1)[1].strip()
                    
                    duty_station_element = details_container.find("span", string="Duty Station: ")
                    if duty_station_element and duty_station_element.parent:
                        duty_station = duty_station_element.find_next_sibling(string=True).strip()

                job_data = {
                    "title": title,
                    "company": company,
                    "link": link,
                    "category": category,
                    "posted_date": posted_date,
                    "deadline_date": deadline_date,
                    "duty_station": duty_station
                }
                
                # Avoid adding duplicate jobs by checking the link
                if not any(existing_job['link'] == link for existing_job in all_jobs):
                    all_jobs.append(job_data)
                else:
                    print(f"Skipped duplicate job: {title}")
                
                if company != "N/A":
                    company_names.add(company)

        print(f"Total jobs collected: {len(all_jobs)}")
        # Filter jobs based on relevance to software development
        if keyword:
            # If a specific keyword is provided, filter by that keyword
            pattern = r'\b' + re.escape(keyword) + r'\b'
            filtered_jobs = [
                job for job in all_jobs 
                if re.search(pattern, job['title'], re.IGNORECASE) or 
                   re.search(pattern, job['category'], re.IGNORECASE)
            ]
        else:
            # When no specific keyword is given, filter using DEFAULT_KEYWORDS
            # for additional relevance check on job titles and descriptions
            keyword_pattern = r'\b(' + '|'.join(DEFAULT_KEYWORDS) + r')\b'
            filtered_jobs = []
            
            for job in all_jobs:
                # Check if the job title or category contains relevant keywords
                if re.search(keyword_pattern, job['title'], re.IGNORECASE):
                    filtered_jobs.append(job)

        return {
            "total_jobs": len(all_jobs),
            "filtered_jobs": len(filtered_jobs),
            "unique_companies": len(company_names),
            "jobs": filtered_jobs
        }

File: job-scraper-backend/src/scrapers/greatrwandajobsScraper.js
// src/scrapers/greatrwandajobsScraper.js

import axios from 'axios';
import * as cheerio from 'cheerio';
import { BaseScraper } from './baseScraper.js';
import { DEFAULT_KEYWORDS } from './keywords.js';

export class GreatRwandaJobsScraper extends BaseScraper {
  /**
   * Scraper for 'greatrwandajobs.com' job listings.
   */

  _isDeadlineValid(deadlineDateStr) {
    if (deadlineDateStr === 'N/A' || !deadlineDateStr.trim()) {
      return true; // Include jobs without deadline information
    }

    try {
      const dateFormats = [
        // "Wednesday, May 28 2025" — full weekday, full month
        /^[A-Za-z]+,\s+([A-Za-z]+)\s+(\d{1,2})\s+(\d{4})$/,
        // "15 July 2025"
        /^(\d{1,2})\s+([A-Za-z]+)\s+(\d{4})$/,
        // "15-07-2025"
        /^(\d{1,2})-(\d{1,2})-(\d{4})$/,
        // "15/07/2025"
        /^(\d{1,2})\/(\d{1,2})\/(\d{4})$/,
        // "2025-07-15"
        /^(\d{4})-(\d{1,2})-(\d{1,2})$/,
        // "July 15, 2025"
        /^([A-Za-z]+)\s+(\d{1,2}),\s+(\d{4})$/,
      ];

      let deadlineDate = null;
      const str = deadlineDateStr.trim();

      // Try Date.parse first (handles many formats)
      const parsed = Date.parse(str);
      if (!isNaN(parsed)) {
        deadlineDate = new Date(parsed);
      }

      if (!deadlineDate) {
        // Try dd-mm-yyyy
        const ddmmyyyy = str.match(/^(\d{1,2})-(\d{1,2})-(\d{4})$/);
        if (ddmmyyyy) {
          deadlineDate = new Date(
            parseInt(ddmmyyyy[3]),
            parseInt(ddmmyyyy[2]) - 1,
            parseInt(ddmmyyyy[1])
          );
        }
      }

      if (!deadlineDate) {
        // Try dd/mm/yyyy
        const ddmmyyyySlash = str.match(/^(\d{1,2})\/(\d{1,2})\/(\d{4})$/);
        if (ddmmyyyySlash) {
          deadlineDate = new Date(
            parseInt(ddmmyyyySlash[3]),
            parseInt(ddmmyyyySlash[2]) - 1,
            parseInt(ddmmyyyySlash[1])
          );
        }
      }

      if (!deadlineDate || isNaN(deadlineDate.getTime())) {
        console.log(`Warning: Could not parse date format: ${deadlineDateStr}`);
        return true; // Include jobs with unparseable dates
      }

      const today = new Date();
      const twoWeeksAgo = new Date(today.getTime() - 14 * 24 * 60 * 60 * 1000);

      return deadlineDate >= twoWeeksAgo;
    } catch (err) {
      console.error(`Error parsing deadline date '${deadlineDateStr}':`, err);
      return true;
    }
  }

  async _fetchCategoriesFromWebsite() {
    const HEADERS = {
      'User-Agent':
        'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/91.0.4472.124 Safari/537.36',
    };

    try {
      const response = await axios.get('https://www.greatrwandajobs.com/jobs/', {
        headers: HEADERS,
        timeout: 20000,
      });

      const $ = cheerio.load(response.data);
      const categorySelect = $('select#category[name="category[]"]');

      if (!categorySelect.length) {
        console.log('Warning: Category select element not found on webpage');
        return [];
      }

      const categories = [];
      categorySelect.find('option').each((_, option) => {
        const value = $(option).attr('value');
        const name = $(option).text().trim();
        if (value && name) {
          categories.push({ value, name });
        }
      });

      console.log(`Successfully fetched ${categories.length} categories from website`);
      return categories;
    } catch (err) {
      console.error(`Error fetching categories from website: ${err.message}`);
      return [];
    }
  }

  async _getAllCategoryUrls() {
    let allCategories = await this._fetchCategoriesFromWebsite();

    if (allCategories.length === 0) {
      console.log('Using fallback static categories');
      allCategories = [
        { value: '10', name: 'Engineering jobs in Rwanda' },
        { value: '13', name: 'Computer/ IT jobs in Rwanda' },
        { value: '47', name: 'Data, Monitoring, and Research jobs in Rwanda' },
        { value: '52', name: 'Technician jobs in Rwanda' },
      ];
    }

    const urls = [];

    for (const category of allCategories) {
      let formattedName = category.name.toLowerCase();
      formattedName = formattedName.replace(/[^a-z0-9\s/\-]/g, '');
      formattedName = formattedName.replace(/[\s/]+/g, '-');
      formattedName = formattedName.replace(/^-+|-+$/g, '');

      const url = `https://www.greatrwandajobs.com/job-categories/newest-jobs/category-${formattedName}-${category.value}`;
      urls.push(url);
    }

    console.log(`Built ${urls.length} category URLs for full fetch`);

    if (urls.length === 0) {
      urls.push(
        'https://www.greatrwandajobs.com/job-categories/newest-jobs/category-computer-it-jobs-in-rwanda-13'
      );
    }

    return urls;
  }

  async fetchAll() {
    const BASE_URL = 'https://www.greatrwandajobs.com';
    const HEADERS = {
      'User-Agent':
        'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/91.0.4472.124 Safari/537.36',
    };

    const allJobs = [];
    const companyNames = new Set();

    const urlsToScrape = await this._getAllCategoryUrls();

    for (const URL of urlsToScrape) {
      console.log(`Scraping URL: ${URL}`);
      let response;
      try {
        response = await axios.get(URL, { headers: HEADERS, timeout: 20000 });
      } catch (err) {
        console.error(`Error fetching the URL ${URL}: ${err.message}`);
        continue;
      }

      const $ = cheerio.load(response.data);
      const jobContainers = $('div#js-jobs-wrapper');
      console.log(`Found ${jobContainers.length} job containers on ${URL}`);

      const jobElements = [];
      jobContainers.each((_, container) => {
        const toprow = $(container).find('div.js-toprow');
        if (toprow.length) {
          jobElements.push(toprow);
        }
      });

      console.log(`Found ${jobElements.length} job elements on ${URL}`);

      for (const jobEl of jobElements) {
        const $job = $(jobEl);
        const titleElement = $job.find('a.jobtitle');
        if (!titleElement.length) {
          console.log('Skipping job element: no title found');
          continue;
        }

        const title = titleElement.text().trim();
        const link = BASE_URL + titleElement.attr('href');

        // Extract company
        let company = 'N/A';
        const companyDiv = $job.find('div.js-image');
        if (companyDiv.length) {
          const companyImg = companyDiv.find('img');
          if (companyImg.length) {
            company = companyImg.attr('title') || 'N/A';
          }
        }

        const detailsContainer = $job.find('div.js-second-row');
        let category = 'N/A';
        let postedDate = 'N/A';
        let deadlineDate = 'N/A';
        let dutyStation = 'N/A';

        if (detailsContainer.length) {
          // Extract category
          const categorySpan = detailsContainer.find('span').filter(function () {
            return $(this).text().trim() === 'Job Category:';
          });
          if (categorySpan.length) {
            const nextText = categorySpan[0].nextSibling;
            if (nextText) {
              category = nextText.nodeValue ? nextText.nodeValue.trim() : 'N/A';
            }
          }

          // Extract posted date
          const postedSpan = detailsContainer.find('span').filter(function () {
            return $(this).text().trim() === 'Posted:';
          });
          if (postedSpan.length) {
            const nextText = postedSpan[0].nextSibling;
            if (nextText) {
              postedDate = nextText.nodeValue ? nextText.nodeValue.trim() : 'N/A';
            }
          }

          // Extract deadline
          const deadlineSpan = detailsContainer.find('span').filter(function () {
            return /Deadline of this Job/i.test($(this).text());
          });
          if (deadlineSpan.length) {
            const deadlineText = deadlineSpan.parent().text().trim();
            if (deadlineText.includes(':')) {
              deadlineDate = deadlineText.split(':').slice(1).join(':').trim();
            }
          }

          // Extract duty station
          const dutySpan = detailsContainer.find('span').filter(function () {
            return $(this).text().trim() === 'Duty Station:';
          });
          if (dutySpan.length) {
            const nextText = dutySpan[0].nextSibling;
            if (nextText) {
              dutyStation = nextText.nodeValue ? nextText.nodeValue.trim() : 'N/A';
            }
          }
        }

        const jobData = {
          title,
          company,
          link,
          category,
          posted_date: postedDate,
          deadline_date: deadlineDate,
          duty_station: dutyStation,
        };

        // Avoid duplicates
        if (!allJobs.some((existing) => existing.link === link)) {
          allJobs.push(jobData);
        } else {
          console.log(`Skipped duplicate job: ${title}`);
        }

        if (company !== 'N/A') {
          companyNames.add(company);
        }
      }
    }

    console.log(`Total jobs collected: ${allJobs.length}`);

    return { allJobs, companyNames: [...companyNames] };
  }

  /**
   * Override filterJobs to also check deadline validity and category field.
   */
  filterJobs(allJobs, keyword = null, customKeywords = null) {
    const filterKeywords = customKeywords || DEFAULT_KEYWORDS;
    let filteredJobs;
    if (keyword) {
      const pattern = new RegExp(
        `\\b${keyword.replace(/[.*+?^${}()|[\]\\]/g, '\\$&')}\\b`,
        'i'
      );
      filteredJobs = allJobs.filter(
        (job) =>
          (pattern.test(job.title) || pattern.test(job.category)) &&
          this._isDeadlineValid(job.deadline_date)
      );
    } else {
      const keywordPattern = new RegExp(`\\b(${filterKeywords.join('|')})\\b`, 'i');
      filteredJobs = allJobs.filter(
        (job) =>
          keywordPattern.test(job.title) && this._isDeadlineValid(job.deadline_date)
      );
    }

    console.log(`Jobs after keyword filtering: ${filteredJobs.length}`);
    return filteredJobs;
  }
}
